webrender/src/device/gfx.rs

```diff
@@ -2137,8 +2116,35 @@ impl<B: hal::Backend> Device<B> {
         let renderer_name = "TODO renderer name".to_owned();
         let features = adapter.physical_device.features();
 
-        let memory_types = adapter.physical_device.memory_properties().memory_types;
-        println!("memory_types: {:?}", memory_types);
+        let memory_properties = adapter.physical_device.memory_properties();
+        let mut heaps = {
+            let types = memory_properties.
+                memory_types
+                .iter()
+                .map(|ref mt| {
+                    let mut config = heaps_config;
+                    if let Some(mut linear) = config.linear {
+                        if !mt.properties.contains(gfx_hal::memory::Properties::CPU_VISIBLE) {
+                            config.linear = None;
+                        } else {
+                            linear.linear_size = linear.linear_size
+                                .min((memory_properties.memory_heaps[mt.heap_index as usize] / 8 - 1).next_power_of_two());
+                        }
+                    }
+                    if let Some(mut dynamic) = config.dynamic {
+                        dynamic.max_block_size = dynamic.max_block_size
+                            .min((memory_properties.memory_heaps[mt.heap_index as usize] / 32 - 1).next_power_of_two());
+                        dynamic.block_size_granularity = dynamic.block_size_granularity
+                            .min((memory_properties.memory_heaps[mt.heap_index as usize] / 1024 - 1).next_power_of_two());
```

Comment:
where do these magic constants (8,32,1024) come from?

Reply:
I used the same values which rendy has here https://github.com/omni-viral/rendy/blob/c1b7d10b32a00eb8a03b1bb02fdd891a800751a4/factory/src/config.rs#L148. I don't know if I should extend the config, and add these limitations as well, or just check if the values are below the maximum limit?

Reply:
Rendy uses this values to limit few constants in `Default` implementation for factory config.
They are bit magic but their intent is clear - make config with too big numbers usable in the system.